File: customer-service/src/main/java/com/wks/servicemarketplace/customerservice/config/DefaultApplicationEventListener.java
package com.wks.servicemarketplace.customerservice.config;

import com.wks.servicemarketplace.customerservice.config.schedulers.Schedulers;
import org.glassfish.jersey.server.monitoring.ApplicationEvent;
import org.glassfish.jersey.server.monitoring.ApplicationEventListener;
import org.glassfish.jersey.server.monitoring.RequestEvent;
import org.glassfish.jersey.server.monitoring.RequestEventListener;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import javax.inject.Inject;

public class DefaultApplicationEventListener implements ApplicationEventListener {

    private final Logger LOGGER = LoggerFactory.getLogger(DefaultApplicationEventListener.class);
    private final Schedulers schedulers;
    private final DatabaseMigration migration;

    @Inject
    public DefaultApplicationEventListener(
            DatabaseMigration migrations,
            Schedulers schedulers
    ) {
        this.migration = migrations;
        this.schedulers = schedulers;
    }

    @Override
    public void onEvent(ApplicationEvent event) {
        switch (event.getType()) {
            case INITIALIZATION_START:
                this.migration.migrate()
                        .mapErr(e -> {
                            LOGGER.error(e.toString());
                            return e;
                        })
                        .expect("Migration Failed");
                break;
            case INITIALIZATION_FINISHED:
                this.schedulers.start();
                break;
            case DESTROY_FINISHED:
                this.schedulers.shutdown();
                break;
        }
    }

    @Override
    public RequestEventListener onRequest(RequestEvent requestEvent) {
        return null;
    }
}

File: service-provider-service/src/main/kotlin/com/wks/servicemarketplace/serviceproviderservice/config/DefaultApplicationEventListener.kt
package com.wks.servicemarketplace.serviceproviderservice.config

import org.glassfish.jersey.server.monitoring.ApplicationEvent
import org.glassfish.jersey.server.monitoring.ApplicationEventListener
import org.glassfish.jersey.server.monitoring.RequestEvent
import org.quartz.Scheduler
import javax.inject.Inject

class DefaultApplicationEventListener @Inject constructor(
    private val scheduler: Scheduler,
    private val migration: DatabaseMigration
) : ApplicationEventListener {
    override fun onEvent(event: ApplicationEvent) {
        when (event.type) {
            ApplicationEvent.Type.INITIALIZATION_START -> migration.migrate()
            ApplicationEvent.Type.INITIALIZATION_FINISHED -> scheduler.start()
            ApplicationEvent.Type.DESTROY_FINISHED -> scheduler.shutdown()
        }
    }

    override fun onRequest(requestEvent: RequestEvent?) = null
}